arrow-fx/src/main/kotlin/arrow/fx/IO.kt

```diff
@@ -994,6 +1046,64 @@ fun <E, A, B, E2 : E> IOOf<E, A>.flatMap(f: (A) -> IOOf<E2, B>): IO<E2, B> =
     else -> IO.Bind(current, f)
   }
 
+/**
+ * Flatten an [IO] with a success of [Either] into an [IO] of [A], when the left is the same error ([E]) as the original [IO] and the right is [A].
+ *
+ * @returns success if [Either.Right] or raises [E] otherwise.
+ *
+ *  * ```kotlin:ank:playground
+ * import arrow.fx.IO
+ * import arrow.fx.unsafeRunSync
+ * import arrow.core.Either
+ *
+ * fun main(args: Array<String>) {
+ *   val result =
+ *   //sampleStart
+ *   IO.just(Either.right("Hello")).flatten().flatMap { IO { "$it World" } }
+ *   //sampleEnd
+ *   println(result.unsafeRunSync())
+ * }
+ * ```
+ */
+fun <E, A> IO<E, EitherOf<E, A>>.flattenEither(): IO<E, A> =
+  flatMap { it.fix().fold(::RaiseError, ::Pure) }
+
+/**
+ * Transform the value of an [IO] into an [Either] and consequently flatten into an [IO]
+ *
+ * @return success when [f] results in [Either.Right] or error when [f] results in [Either.Left]
+ *
+ * fun main(args: Array<String>) {
+ *   fun Int.increment() = Either.right(this + 1)
+ *   val result =
+ *   //sampleStart
+ *   IO.just(1).mapEither { it.increment() }
+ *   //sampleEnd
+ *   println(result.unsafeRunSyncEither())
+ * }
+ * ```
+ */
+fun <E, A, B> IO<E, A>.mapEither(f: (A) -> EitherOf<E, B>): IO<E, B> =
```

Comment:
If I'm not mistaken, this function should already be covered by `effectMapEither` because you can still pass a non-suspend function to that one. So I'd personally remove this one, no? cc @pakoito @nomisRev

Reply:
It does work, however, we may not force a suspension unless necessary. The bytecode is quite different:
For:
```kotlin
val run = IO.just(1).mapEither { Either.right(it + 1) }
```
it generates the equivalent to:
```Java
run = IOKt.mapEither(IO.Companion.just(1), (Function1)null.INSTANCE);
```
On the other hand, if we use suspension:
```kotlin
val runAsync = IO.just(1).effectMapEither { Either.right(it + 1) }
```
it generates the equivalent to:
```java
     runAsync = IOKt.effectMapEither(IO.Companion.just(1), (Function2)(new Function2((Continuation)null) {
         private int p$0;
         int label;

         @Nullable
         public final Object invokeSuspend(@NotNull Object $result) {
            Object var3 = IntrinsicsKt.getCOROUTINE_SUSPENDED();
            switch(this.label) {
            case 0:
               ResultKt.throwOnFailure($result);
               int it = this.p$0;
               return Either.Companion.right(Boxing.boxInt(it + 1));
            default:
               throw new IllegalStateException("call to 'resume' before 'invoke' with coroutine");
            }
         }

         @NotNull
         public final Continuation create(@Nullable Object value, @NotNull Continuation completion) {
            Intrinsics.checkParameterIsNotNull(completion, "completion");
            Function2 var3 = new <anonymous constructor>(completion);
            ((Number)value).intValue();
            var3.p$0 = ((Number)value).intValue();
            return var3;
         }

         public final Object invoke(Object var1, Object var2) {
            return ((<undefinedtype>)this.create(var1, (Continuation)var2)).invokeSuspend(Unit.INSTANCE);
         }
      }));
```

Reply:
Yep, it's a perf thing. We want `map` separate from `effectMap`.